File: protected/controllers/SpatialDataController.vb
Imports GeoPortal.GeoPortal.Models.Data
Imports GeoPortal.GeoPortal.Models.BusinessLogic
Imports Newtonsoft.Json.Linq
Imports System.Xml
Imports System.IO
Imports System.Drawing

Namespace GeoPortal
    Public Class SpatialDataController
        Inherits System.Web.Mvc.Controller

        '
        ' GET: /SpatialData
        <CompressFilter()>
        Function getSpatialUnits(ByVal surveyID As String) As String

            Dim SD As New SpatialData

            Dim SU As ArrayList = SD.getAvailableUnits(surveyID)

            Return "{""rows"":" & Newtonsoft.Json.JsonConvert.SerializeObject(SU) & "}"


        End Function
        <CompressFilter()>
        Function getSpatialLabel(ByVal TableName As String) As String

            Dim SD As New SpatialData

            Dim SL As ArrayList = SD.getSpatialLabels(TableName)


            Return "{""rows"":" & Newtonsoft.Json.JsonConvert.SerializeObject(SL) & "}"
        End Function
        <CompressFilter()>
        Function getChoroFields(ByVal TableName As String) As String

            Dim SD As New SpatialData

            Dim SL As ArrayList = SD.getChoroFields(TableName)


            Return "{""rows"":" & Newtonsoft.Json.JsonConvert.SerializeObject(SL) & "}"
        End Function

        <CompressFilter()>
        Function getSpatialSubUnit(ByVal TableName As String) As String


            Dim SD As New SpatialData

            Dim SSU As ArrayList = SD.getSpatialSubUnits(TableName)

            Return "{""rows"":" & Newtonsoft.Json.JsonConvert.SerializeObject(SSU) & "}"


        End Function


        <CompressFilter()>
        Function getSpatialRefSubUnit(ByVal SID As String, ByVal MajorUnit As String) As String


            Dim SD As New SpatialData

            Dim RSSU As ArrayList = SD.getRefSpatialSubUnits(getTableName(SID, MajorUnit))

            Return "{""rows"":" & Newtonsoft.Json.JsonConvert.SerializeObject(RSSU) & "}"


        End Function


        <CompressFilter()>
        Function getRefSpatialIndivudalUnits(ByVal UnitName As String, ByVal MajorUnit As String, ByVal SubUnit As String, ByVal the_geom As String, ByVal SID As String) As String


            Dim SD As New SpatialData

            Dim ISSU As ArrayList = SD.getRefSpatialIndivudalUnits(UnitName, MajorUnit, SubUnit, the_geom, SID)

            Return "{""rows"":" & Newtonsoft.Json.JsonConvert.SerializeObject(ISSU) & "}"


        End Function

        Private Function getTableName(ByVal SID As String, ByVal unit As String) As String


            If unit = "Police Region" Then
                Return "x_" & SID & "_police_"
            ElseIf unit = "Assembly Economic Fora Area" Then
                Return "x_" & SID & "_aefa_"
            ElseIf unit = "Fire Brigade Region" Then
                Return "x_" & SID & "_fire_"
            ElseIf unit = "Lower Super Output Area" Then
                Return "x_" & SID & "_lsoa_"
            ElseIf unit = "Parliamentary Constituencies" Then
                Return "x_" & SID & "_parl_"
            ElseIf unit = "Postcode Sector" Then
                Return "x_" & SID & "_pcode_"
            ElseIf unit = "Unitary Authority" Then
                Return "x_" & SID & "_ua_"
            End If




        End Function


        <CompressFilter()>
        Public Function GenerateSpatialData(ByVal SurveyID As String, ByVal Unit As String, ByVal SubUnit As String, ByVal Outline As String, ByVal Label As String, ByVal fromColour As String, ByVal toColour As String, ByVal Choropleth As String, ByVal ChoroplethField As String, ByVal addLabels As String, ByVal ClassMethod As String, ByVal intervals As Integer) As String

            Dim SD As New SpatialData



            Dim shapes As ArrayList = SD.GenerateSpatialData(SurveyID, Unit, SubUnit, GetBool(Unit), Label, fromColour, toColour, GetBool(Choropleth), ChoroplethField, GetBool(addLabels), ClassMethod, intervals)

            If shapes.Count > 0 Then
                Return "({""success"": true, ""shapes"": " & Newtonsoft.Json.JsonConvert.SerializeObject(shapes) & "})"
            Else
                Dim msg As New GeoPortal.Models.BusinessLogic.jsonMsg
                msg.success = False
                msg.message = "No Spatial Data currently available for this record."
                Return Newtonsoft.Json.JsonConvert.SerializeObject(msg)
            End If




        End Function


        <CompressFilter()>
        Public Function getSpatialDataSets(ByVal data As String)

            Dim dataSet As New Dictionary(Of String, spatialSearchLayers)

            Dim jsons As Newtonsoft.Json.Linq.JArray = Newtonsoft.Json.JsonConvert.DeserializeObject(data)

            For Each json As JToken In jsons

                Dim layer As spatialSearchLayers = Newtonsoft.Json.JsonConvert.DeserializeObject(Of spatialSearchLayers)(json.ToString)
                dataSet.Add(layer.name, layer)
            Next

            Dim spatialSearch As New SpatialData

            Dim full_data As Dictionary(Of String, spatialSearchLayers) = spatialSearch.getChosenLayers(dataSet)

        End Function



        <CompressFilter()>
        Public Function SpatialSearch(ByVal geography As String, ByVal start As Integer, ByVal limit As Integer, ByVal Type As String) As String

            Dim count As Integer

            Dim SD As New SpatialData


            Dim results As ArrayList

            'Return "{""success"": true, ""data"":" & Newtonsoft.Json.JsonConvert.SerializeObject(results) & "}"

            Dim res As GeoPortal.Models.BusinessLogic.SpatialSearch2

            If geography = "" Then
                results = Session("spatialResults")
                Select Case Type
                    Case "Quant"
                        count = Session("quantCount")
                    Case "Qual"
                        count = Session("qualCount")
                    Case "Grey"

                    Case "Admin"

                End Select

                res = results(0)
            Else
                results = Nothing
                If results Is Nothing Then
                    results = SD.SpatialSearch(geography)

                    res = results(0)
                    Select Case Type
                        Case "Quant"
                            count = res.quantCount
                            Session.Add("quantCount", count)
                        Case "Qual"
                            count = res.qualCount
                            Session.Add("qualCount", count)
                        Case "Grey"

                        Case "Admin"

                    End Select

                    'Return "{""quantData"":" & Newtonsoft.Json.JsonConvert.SerializeObject(res.quantData) & "}"

                    Session.Add("spatialResults", results)

                End If
            End If


            'Dim resultsset As New GeoPortal.Models.BusinessLogic.spatialResults

            'resultsset.totalCount = count
            'resultsset.questions = Newtonsoft.Json.JsonConvert.SerializeObject(results(0).quantData)


            Dim pageResults As New ArrayList

            Dim cnt As Integer = start
            Dim cnt_end As Integer = cnt + limit

            Dim Str As String = ""

            Select Case Type
                Case "Quant"

                    Do Until cnt = cnt_end Or cnt = res.quantCount

                        pageResults.Add(res.quantData(res.quantData.Keys(cnt)))
                        cnt += 1
                    Loop

                    Str = "{""totalCount"":" & count.ToString & ",""quantData"":" & Newtonsoft.Json.JsonConvert.SerializeObject(pageResults) & "}"


                Case "Qual"
                    Do Until cnt = cnt_end Or cnt = res.qualCount

                        pageResults.Add(res.qualData(res.qualData.Keys(cnt)))
                        cnt += 1
                    Loop

                    Str = "{""totalCount"":" & count.ToString & ",""qualData"":" & Newtonsoft.Json.JsonConvert.SerializeObject(pageResults) & "}"
                Case "Grey"

                Case "Admin"

            End Select

            Return Str

        End Function


        <CompressFilter()>
        Public Function VerifySpatialSearch(ByVal coords As String, ByVal type As String, ByVal dist As Integer) As String

            Dim SD As New SpatialData
            Dim msg As New jsonMsg
            If SD.VerifySpatialSearch(coords, type, dist) Then
                msg.success = True
                msg.message = "Results Found"
            Else
                msg.success = False
                msg.message = "No Results found within search distance"

            End If



            Return Newtonsoft.Json.JsonConvert.SerializeObject(msg)


        End Function

        <CompressFilter()>
        Public Function GenerateQualSpatialData(ByVal ID As String) As String

            Dim SD As New SpatialData
            Dim shapes As ArrayList = SD.generateQualSpatialData("red", ID)

            If shapes.Count > 0 Then
                Return "({""success"": true, ""shapes"": " & Newtonsoft.Json.JsonConvert.SerializeObject(shapes) & "})"
            Else
                Dim msg As New GeoPortal.Models.BusinessLogic.jsonMsg
                msg.success = False
                msg.message = "No Spatial Data currently available for this record."
                Return Newtonsoft.Json.JsonConvert.SerializeObject(msg)
            End If

        End Function


        Private Function GetBool(ByVal Status As String) As Boolean

            If Status = "on" Then
                Return True
            Else
                Return False
            End If

        End Function


        Public Function getMinMax(ByVal TableName As String, ByVal ColName As String) As String

            Dim getMin As String = "Select MIN(" & ColName & ") from " & TableName
            Dim getMax As String = "Select MAX(" & ColName & ") from " & TableName

            Dim db As New getDBConnections()

            Dim dBCnn As Npgsql.NpgsqlConnection = db.getDBConnection("Survey_Data")
            Dim cmd As New Npgsql.NpgsqlCommand(getMin, dBCnn)

            dBCnn.Open()

            Dim min As Integer = cmd.ExecuteScalar()

            cmd.CommandText = getMax

            Dim max As Integer = cmd.ExecuteScalar()

            dBCnn.Close()


            Dim minMax(1) As Integer
            minMax(0) = min
            minMax(1) = max

            Return Newtonsoft.Json.JsonConvert.SerializeObject(minMax).ToString

        End Function

        <CompressFilter()>
        Public Function DynamicSLD2(ByVal type As String, ByVal fromColour As String, ByVal toColour As String, ByVal fieldName As String, ByVal min As Integer, ByVal max As Integer, ByVal classes As Integer, ByVal layer As String, ByVal labelName As String) As String

            Dim SLD As New SLD

            Dim xmlDoc As XmlDocument = SLD.DynamicSLD2(type, fromColour, toColour, fieldName, min, max, classes, layer, labelName)

            Dim sw As New StringWriter()
            Dim txml As New XmlTextWriter(sw)
            xmlDoc.WriteTo(txml)


            Return sw.ToString
        End Function




        <CompressFilter()>
        Public Function getFeatureInfo(ByVal lat As Double, lon As Double, tableIDs As String) As String

            ' create a list of tables from the JSON 
            Dim tables As List(Of IdentifyFeatures) = Newtonsoft.Json.JsonConvert.DeserializeObject(Of List(Of IdentifyFeatures))(tableIDs)
            Dim SD As New SpatialData
            Dim results As ArrayList = SD.getFeatureInfoTable(lat, lon, tables)


            Return Newtonsoft.Json.JsonConvert.SerializeObject(results)


        End Function

    End Class
End Namespace

File: protected/controllers/SpatialDataController.php
﻿<?
class SpatialDataController extends Controller {

    // '
    // ' GET: /SpatialData
    // // <CompressFilter()>
    Function actiongetSpatialUnits() {

        $surveyID = "";
        if(isset($_POST['SurveyID'])) {
            $surveyID = $_POST['SurveyID'];
        }

        $SD = New SpatialData();

        $SU = $SD->getAvailableUnits($surveyID);

        echo '{"rows":' . json_encode($SU) . "}";


    }
    // // <CompressFilter()>
    Function actiongetSpatialLabel($TableName) {

        $SD = New SpatialData();

        $SL = $SD->getSpatialLabels($TableName);


        echo '{"rows":' . json_encode($SL) . "}";
    }
    // // <CompressFilter()>
    Function actiongetChoroFields($TableName) {

        $SD = New SpatialData();

        $SL = $SD->getChoroFields($TableName);


        echo '{"rows":' . json_encode($SL) . "}";
    }

    // // <CompressFilter()>
    Function actiongetSpatialSubUnit($TableName) {


        $SD = New SpatialData();

        $SSU = $SD->getSpatialSubUnits($TableName);

        echo '{"rows":' . json_encode($SSU) . "}";


    }


    // // <CompressFilter()>
    Function actiongetSpatialRefSubUnit($SID, $MajorUnit) {


        $SD = New SpatialData();

        $RSSU = $SD->getRefSpatialSubUnits($this::getTableName($SID, $MajorUnit));

        echo '{"rows":' . json_encode($RSSU) . "}";


    }


    // // <CompressFilter()>
    Function actiongetRefSpatialIndivudalUnits($UnitName, $MajorUnit, $SubUnit, $the_geom, $SID) {


        $SD = New SpatialData();

        $ISSU = $SD->getRefSpatialIndivudalUnits($UnitName, $MajorUnit, $SubUnit, $the_geom, $SID);

        echo '{"rows":' . json_encode($ISSU) . "}";


    }

    Private Function getTableName($SID, $unit) {

        If ( $unit == "Police Region" ) {
            Return "x_" . $SID . "_police_";
        } ElseIf ( $unit == "Assembly Economic Fora Area" ) {
            Return "x_" . $SID . "_aefa_";
        } ElseIf ( $unit == "Fire Brigade Region" ) {
            Return "x_" . $SID . "_fire_";
        } ElseIf ( $unit == "Lower Super Output Area" ) {
            Return "x_" . $SID . "_lsoa_";
        } ElseIf ( $unit == "Parliamentary Constituencies" ) {
            Return "x_" . $SID . "_parl_";
        } ElseIf ( $unit == "Postcode Sector" ) {
            Return "x_" . $SID . "_pcode_";
        } ElseIf ( $unit == "Unitary Authority" ) {
            Return "x_" . $SID . "_ua_";
        }

    }


    // // <CompressFilter()>
    Public Function actionGenerateSpatialData($SurveyID, $Unit, $SubUnit, $Outline, $Label, $fromColour, $toColour, $Choropleth, $ChoroplethField, $addLabels, $ClassMethod, $intervals) {

        $SD = New SpatialData();

        $shapes = $SD->GenerateSpatialData($SurveyID, $Unit, $SubUnit, GetBool($Unit), $Label, $fromColour, $toColour, GetBool($Choropleth), $ChoroplethField, GetBool($addLabels), $ClassMethod, $intervals);

        If ( sizeof($shapes) > 0 ) {
            Return '({"success": true, "shapes": ' . json_encode($shapes) . "})";
        } Else {
            $msg = New jsonMsg();
            $msg->success = False;
            $msg->message = "No Spatial Data currently available for this record.";
            Return json_encode($msg);
        }

    }


    // // <CompressFilter()>
    Public Function actiongetSpatialDataSets($data) {

        $dataSet = array();

        $jsons = json_decode($data, true); //Newtonsoft.Json.Linq.JArray

        ForEach ($jsons as $json) {

            $layer = json_decode($json);
            $dataSet[$layer->name] = $layer;
        }

        $spatialSearch = New SpatialData();

        $full_data = $spatialSearch->getChosenLayers($dataSet);

    }



    // <CompressFilter()>
    Public Function actionSpatialSearch() {

//        $count;
        $geography = "";
        if(isset($_POST['geography'])) {
            $geography = $_POST['geography'];
        }
        $start = "";
        if(isset($_POST['start'])) {
            $start = $_POST['start'];
        }
        $limit = "";
        if(isset($_POST['limit'])) {
            $limit = $_POST['limit'];
        }
        $Type = "";
        if(isset($_POST['type'])) {
            $Type = $_POST['type'];
        }

        $SD = New SpatialData();


        $results = array();

        $count = 0;

        // 'Return "{""success"": true, ""data"":" . json_encode($results) . "}";

        $res = new SpatialSearch2();

        If ( $geography == "" ) {
            $results = Yii::app()->session["spatialResults"];
            Switch ($Type) {
                case "Quant":
                    $count = intval(Yii::app()->session["quantCount"]);
                    break;
                case "Qual":
                    $count = intval(Yii::app()->session["qualCount"]);
                    break;
                case "Grey":
                    break;
                case "Admin":
                    break;
            }

            $res = $results[0];
        } Else {
            $results = null;
            If ( $results == null ) {
                $results = $SD->SpatialSearch($geography);

                $res = $results[0];
                Switch ($Type) {
                    case "Quant":
                        $count = intval($res->quantCount);
                        Yii::app()->session["quantCount"] = $count;
                        break;
                    case "Qual":
                        $count = intval($res->qualCount);
                        Yii::app()->session["qualCount"] = $count;
                        break;
                    case "Grey":

                        break;
                    case "Admin":

                }

                // 'Return "{""quantData"":" . json_encode(res.quantData) . "}";

                Yii::app()->session["spatialResults"] = $results;

            }
        }


        // '$resultsset = New GeoPortal.Models.BusinessLogic.spatialResults

        // 'resultsset.totalCount = $count
        // 'resultsset.questions = json_encode(results(0).quantData);


        $pageResults = array();

        $cnt = intval($start);
        $cnt_end = intVal($cnt) + intVal($limit);

        $Str = "";

        Switch ($Type) {

            case "Quant":

//                Log::toFile('cnt_end and quantCount : ' . $cnt_end . ' : ' . $res->quantCount);
//                Log::toFile('quantData : ' . print_r($res->quantData, true));
                $keys = array_keys($res->quantData);

                while ( $cnt < $cnt_end && $cnt < $res->quantCount) {

                    $pageResults[] = ($res->quantData[$keys[$cnt]]);
                    $cnt ++;
                }

//                Log::toFile('quantResults ' . print_r($pageResults, true));

                $Str = '{"totalCount":' . $count . ',"quantData":' . json_encode($pageResults) . "}";


                break;
            case "Qual":

//                Log::toFile('cnt_end and qualCount : ' . $cnt_end . ' : ' . $res->quantCount);
//                Log::toFile('qualData : ' . print_r($res->qualData, true));
                $keys = array_keys($res->qualData);

                while( $cnt < $cnt_end && $cnt < $res->qualCount) {

                    $pageResults[] = ($res->qualData[$keys[$cnt]]);
                    $cnt ++;
                }
//                Log::toFile('quantResults ' . print_r($pageResults, true));

                $Str = '{"totalCount":' . $count . ',"qualData":' . json_encode($pageResults) . '}';
                break;
            case "Grey":

                break;
            case "Admin":
                break;
        }

        echo $Str;

    }


    // <CompressFilter()>
    Public Function actionVerifySpatialSearch($coords, $type, $dist) {

        $SD = New SpatialData();
        $msg = New jsonMsg();
        If ( $SD->VerifySpatialSearch($coords, $type, $dist) ) {
            $msg->success = True;
            $msg->message = "Results Found";
        } Else {
            $msg->success = False;
            $msg->message = "No Results found within search distance";

        }



        Return json_encode($msg);


    }

    // <CompressFilter()>
    Public Function actionGenerateQualSpatialData() {

        $ID = "";
        if(isset($_POST['ID'])) {
            $ID = $_POST['ID'];
        }

        $SD = New SpatialData();
        $shapes = $SD->generateQualSpatialData("red", $ID);

        If ( sizeof($shapes) > 0 ) {
            echo '({"success": true, "shapes": ' . json_encode($shapes) . "})";
        } Else {
            $msg = New jsonMsg();
            $msg->success = False;
            $msg->message = "No Spatial Data currently available for this record.";
            echo json_encode($msg);
        }

    }


    Private Function actionGetBool($Status) {
        If ( $Status == "on" ) {
            Return True;
        } Else {
            Return False;
        }

    }


    Public Function actiongetMinMax($TableName, $ColName) {

        $getMin = "Select MIN(" . $ColName . ") from " . $TableName;
        $getMax = "Select MAX(" . $ColName . ") from " . $TableName;

        $db = New getDBConnections();

        $dBCnn = $db->getDBConnection("Survey_Data");
        $cmd = pg_query($dBCnn, $getMin);

//            dBCnn.Open();

        $min = $cmd->ExecuteScalar();

        $cmd->CommandText = $getMax;

        $max = $cmd->ExecuteScalar();

//            dBCnn.Close();


        $minMax = array();
        $minMax[0] = $min;
        $minMax[1] = $max;

        Return json_encode($minMax);

    }

    // <CompressFilter()>
    Public Function actionDynamicSLD() {
        $type = "";
        if(isset($_POST['type'])) {
            $type = $_POST['type'];
        }
        $fromColour = "";
        if(isset($_POST['fromColour'])) {
            $fromColour = $_POST['fromColour'];
        }
        $toColour = "";
        if(isset($_POST['toColour'])) {
            $toColour = $_POST['toColour'];
        }
        $fieldName = "";
        if(isset($_POST['fieldName'])) {
            $fieldName = $_POST['fieldName'];
        }
        $min = "";
        if(isset($_POST['min'])) {
            $min = $_POST['min'];
        }
        $max = "";
        if(isset($_POST['max'])) {
            $max = $_POST['max'];
        }
        $classes = "";
        if(isset($_POST['classes'])) {
            $classes = $_POST['classes'];
        }
        $layer = "";
        if(isset($_POST['layer'])) {
            $layer = $_POST['layer'];
        }
        $labelName = "";
        if(isset($_POST['labelName'])) {
            $labelName = $_POST['labelName'];
        }

        $SLD = New SLD();

        $xmlDoc = $SLD->DynamicSLD2($type, $fromColour, $toColour, $fieldName, $min, $max, $classes, $layer, $labelName);
//
//        $sw = New StringWriter();
//        $txml = New XmlTextWriter($sw);
//        $xmlDoc->WriteTo($txml);
//

        echo $xmlDoc;
    }




    // <CompressFilter()>
    Public Function actiongetFeatureInfo($lat, $lon, $tableIDs) {

        // ' create a list of tables from the JSON
        $tables = json_decode($tableIDs);
        $SD = New SpatialData();
        $results = $SD->getFeatureInfoTable($lat, $lon, $tables);


        Return json_encode($results);


    }

}